OIDC: Add bearer token step up authentication (#47445)
quarkusio/quarkus: docs/src/main/asciidoc/security-oidc-bearer-token-authentication.adoc

For more complex setup involving multiple tenants please see the xref:security-openid-connect-multitenancy.adoc#programmatic-startup[Programmatic OIDC start-up for multitenant application]
section of the OpenID Connect Multi-Tenancy guide.

== Step Up Authentication

> The `io.quarkus.oidc.AuthenticationContext` annotation can be used to enforce required authentication level for the Jakarta REST resource classes and methods.

Reviewer: ```suggestion
The `io.quarkus.oidc.AuthenticationContext` annotation can be used to list one or more Authentication Context Class Reference (ACR) values to enforce a required authentication level for the Jakarta REST resource classes and methods.
```

> The `io.quarkus.oidc.AuthenticationContext` annotation can be used to enforce required authentication level for the Jakarta REST resource classes and methods.

Reviewer: Also, since it can also be effective for WebSockets Next, may be we should just say `to enforce a required authentication level` and leave it that ? FYI, the annotation itself just says `endpoints` without listing the actual technologies

Author: This is similar situation to what we have for `@Tenant`, I am bit worried to avoid mentioning Jakarta REST and WS Next because users have no means to figure if it works for gRPC / GraphQL / MCP Server etc. (I suppose it doesn't). I got what you mean and maybe it can be better phrased, but I think we need to be explicit.

Author: You can suggest alternative way to say that though.

Reviewer: @michalvavrik Sounds good, let's keep it as is for now

> The https://datatracker.ietf.org/doc/rfc9470/[OAuth 2.0 Step Up Authentication Challenge Protocol] introduces a mechanism for resource servers to request stronger token authentication methods.

Reviewer: ```suggestion
The https://datatracker.ietf.org/doc/rfc9470/[OAuth 2.0 Step Up Authentication Challenge Protocol] introduces a mechanism for resource servers to request stronger authentication methods when the token does not have expected Authentication Context Class Reference (ACR) values.
```

Consider the following example:

[source,java]
----
package io.quarkus.it.oidc;

import io.quarkus.oidc.AuthenticationContext;
import io.quarkus.oidc.BearerTokenAuthentication;
import jakarta.ws.rs.GET;
import jakarta.ws.rs.Path;

@BearerTokenAuthentication
@Path("/")
public class GreetingsResource {

    @Path("hello")
    @AuthenticationContext("myACR") <1>
    @GET
    public String hello() {
        return "hello";
    }

    @Path("hi")

>     @AuthenticationContext(maxAge = 120) <2>

Reviewer: If we have `myACR` requirement at the class level but only `maxAge` at the method level, should we inherit.

How about us making the `value` (acr) property required, and if users want, they can add it time constrained ? That would avoid some possibly tricky situation where a class level acr is set and some sub-resource method has only `maxAge` ?

Author: Done.

    @GET
    public String hi() {
        return "hi";
    }
}
----

> <1> Bearer access token's `acr` claim must include the `myACR` authentication context class.

Reviewer: ```suggestion
<1> Access token must have an `acr` claim with the `myACR` authentication context class reference value.
```

> <2> Bearer access token's expiration time is the `max_age` annotation attribute value added to the `auth_time` token claim value.

Reviewer: ```suggestion
<2> Bearer access token must have an `acr` claim with the `myACR` ACR value and be in use for no longer than 120 minutes since the authentication time represented by the `auth_time` token claim value.
```
The token age limits how long an otherwise valid token can be used for in a specific context, the token's expiry is a different value. You are right it is kind of similar, but it can be a bit confusing talking about the expiry.

I'd consider dropping a reference to the `auth_time` if the spec offers a fallback to the `issued at` (iat) claim, if no `auth_time` is present.

Also, I suggested indirectly to use minutes as seconds is too small a unit for the access tokens and it can be harder to calculate longer token ages

Author: > The token age limits how long an otherwise valid token can be used for in a specific context, the token's expiry is a different value. You are right it is kind of similar, but it can be a bit confusing talking about the expiry.

I didn't realize that, thanks Sergey

> I'd consider dropping a reference to the auth_time if the spec offers a fallback to the issued at (iat) claim, if no auth_time is present.

Specs says `auth_time` must be there if authorization request contained `max_time`. Maybe I don't understand what you mean. Your docs change suggestion is alright. So far _issued at (iat)_ cannot be used until specs changes.

> Also, I suggested indirectly to use minutes as seconds is too small a unit for the access tokens and it can be harder to calculate longer token ages

I'll change that together with switch to the duration.

Reviewer: @michalvavrik Hi Michal, thanks. In general, we do not have to follow specs the latter to latter.
What happens if the user says `maxAge=...` and the token has no `auth_time` ? Spec is one thing but if SPA works with a provider which simply does not set this claim on the token, then SPA can't do Step-Up authentication which would not be great.

Well, in this case, we can log a warning and say, we can't check it, `maxAge` is ignored. Or we can log an info or debug message - the `auth_time` is not available, using `issued at` claim value. `Issued at` is likely going to be nearly the same as `auth_time` and is always going to stay a few milliseconds or a few seconds max earlier than `auth_time`. And if both claims are absent - we fail the request

Author: Alright, I'll add `iat` support.

Author: Done.
